server/tests/unittests/test_user.py

```diff
@@ -39,5 +44,115 @@ class UserPermissionsUnitTest(PermissionsUnitTest):
+        self.admin = models.User(email=['test@admin.com'], is_admin=True)
+        self.admin.put()
+
+        self.course = make_fake_course(self.admin)
+        self.course.put()
+        self.assign = make_fake_assignment(self.course, self.admin)
+        self.assign.put()
+
+    def tearDown(self):
+        super(UserUniquenessTest, self).tearDown()
+        self.assertNoException()
+
+    def check_user_maker(called_func): #pylint: disable=no-self-argument
+        def wrapped(self, user_id):
+            if isinstance(user_id, ndb.Key):
+                user_id = user_id.get()
+
+            try:
+                called_func(user_id)
+            except Exception as e: #pylint: disable=broad-except
+                self._caught_exception = e #pylint: disable=protected-access
+        return wrapped
+
+    check_user_email = check_user_maker(utils.unique_email_address)
+    check_user_final = check_user_maker(utils.unique_group)
+    check_user_group = check_user_maker(utils.unique_final_submission)
+
+    def get_basic_user(self): #pylint: disable=no-self-use
+        email = 'test@example.com'
+        user = models.User(email=[email])
+        user.put()
+
+        return user
+
+    def assertNoException(self):
+        self.assertEqual(
+            self._caught_exception, None, "Unexpected exception: {}".format(
+                self._caught_exception))
+
+    def assertNumGroups(self, user, assignment, num):
+        G = models.Group
+        groups = G.query(
+            G.member == user.key, G.assignment == assignment.key).fetch()
+        self.assertEqual(len(groups), num)
+
+    def assertNumFS(self, user, assignment, num):
+        FS = models.FinalSubmission
+        subms = FS.query(
+            FS.submitter == user.key, FS.assignment == assignment.key).fetch()
+        self.assertEqual(len(subms), num)
+
+    def assertNumUsers(self, user, num):
+        U = models.User
+        found = []
+        for email in user.email:
+            users = U.query(U.email == email).fetch()
+            for u in users:
+                found.append(u)
+
+        self.assertEqual(len(found), num)
+
+    def test_user_unique_email_single_user(self):
+        self.check_user_email(self.user)
+        self.assertNumUsers(self.user, 1)
+        self.assertNumFS(self.user, self.assign, 0)
+        self.assertNumGroups(self.user, self.assign, 0)
+
+    def test_user_unique_group_single_user(self):
+        self.check_user_group(self.user)
+        self.assertNumUsers(self.user, 1)
+        self.assertNumFS(self.user, self.assign, 0)
+        self.assertNumGroups(self.user, self.assign, 0)
+
+    def test_user_unique_final_single_user(self):
+        self.check_user_final(self.user)
+        self.assertNumUsers(self.user, 1)
+        self.assertNumFS(self.user, self.assign, 0)
+        self.assertNumGroups(self.user, self.assign, 0)
+
+    def test_call_functions(self):
+        user_obj = models.User(email=['test@example.com'])
```

Comment:
The idea is that it tests to make sure that these functions are actually called. Not how many times. I can change it to not test how many times it gets called if you want.